Replat blazor on net5

pranavkm opened the pull request:
Still need to get through some test issues, but the apps work now without issues and the sdk tests are passing locally. 

Things pending

* Revive gzip compression (TBD based on discussion)
* Fix skipped tests
* ~~Scramble the dotnet.js file name~~
* ~~Figure out packing for blazor.webassembly.js~~
* ~~Update templates to net5~~

The largest issue I noticed since the sync up today is that `RuntimeIdentifer` is required to be specified in the project file. The Razor SDK is imported too late to set this up correctly. ~~I'm going to talk to the SDK folks to figure out how to not have this in the project file. Ignore that for now.~~ We had a conversation with some folks about possibly having a Blazor WASM specific SDK. This should remedy the problem. I'll hold off on this for now.

pranavkm committed:
Replat blazor on net5

captainsafia commented on src/Components/Components.sln:
```diff
@@ -15,10 +15,6 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Compon
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Components.WebAssembly.Tests", "WebAssembly\WebAssembly\test\Microsoft.AspNetCore.Components.WebAssembly.Tests.csproj", "{958AD6D2-174B-4B5B-BEFC-FA64B5159334}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Components.WebAssembly.Build", "WebAssembly\Build\src\Microsoft.AspNetCore.Components.WebAssembly.Build.csproj", "{E8AD67A4-77D3-4B85-AE19-4711388B62B1}"
```
Will our build tasks be relocated to the `Microsoft.NET.Sdk.Razor` project now?

pranavkm replied:
Yes, that's my goal. We were also talking about having a Blazor WASM specific SDK, but that's still to be figured out.

captainsafia replied:
Got it. I do like the idea of having a Blazor WASM specific SDK. For a beginner, it helps make the distinction between Razor/Blazor Server/Blazor WASM a little bit easier to understand.

pranavkm committed:
Make all e2e tests pass

pranavkm commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -204,11 +204,11 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
 
   // Begin loading the .dll/.pdb/.wasm files, but don't block here. Let other loading processes run in parallel.
   const dotnetWasmResourceName = 'dotnet.wasm';
-  const assembliesBeingLoaded = resourceLoader.loadResources(resources.assembly, filename => `_framework/_bin/${filename}`, 'assembly');
```
Do we care about keeping these paths in tact? It does break users that scripted a custom resource loader, but it's small enough to fix. I'd prefer to not have an intimate understanding of what the runtime pack looks like in our SDK.

captainsafia replied:
I'm OK with removing it. IMO, it makes more sense to resolve the path of the assembly at a higher level and keep this resource loader location-agnostic.

SteveSandersonMS replied:
I'm fine with the paths changing (if they're retaining at least the `_framework` prefix, since that's the bit we've regarded as indicative that the rest is private).

Custom resource loaders are meant not to rely on specific URL formats. That's why in the "load resource" callback we supply the resource type, resource name, and default URL all as separate args. People shouldn't assume the default URL matches any specific format.

pranavkm commented on src/Components/WebAssembly/testassets/MonoSanity/MonoSanity.csproj:
```diff
@@ -1,16 +0,0 @@
-<Project Sdk="Microsoft.NET.Sdk.Web">
```
I didn't bring this app forward. It seems low value and requires additional work to get the loader script functioning. Let me know if you feel strongly about keeping this.

SteveSandersonMS replied:
I think it's great to remove this 👍 

This project (and its tests) was created before Blazor, so we had some way to make sense of how .NET code should behave on WebAssembly, and so that if the runtime behaviors around low-level basics like exception handling, boxing, etc. varied then we'd have a way to observe that objectively.

By now, .NET on WebAssembly is well enough established that we don't have to get independent proof that the low-level basics work correctly!

pranavkm commented on src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Wasm.Authentication.Server.csproj:
```diff
@@ -4,7 +4,6 @@
     <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
     <!-- Avoid CS1705 errors due to mix of assemblies brought in transitively. -->
 
-    <!-- This project references the shared framework transitively. Prevent restore errors by setting this flag. -->
```
Will undo

pranavkm commented on src/Components/test/E2ETest/Tests/WebAssemblyAuthenticationTests.cs:
```diff
@@ -63,13 +63,13 @@ protected override void InitializeAsyncCore()
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/runtime/issues/38098")]
```
Async JSInterop code paths are busted because TCS.SetResult throws

pranavkm commented on src/Components/test/E2ETest/Tests/WebAssemblyLocalizationTest.cs:
```diff
@@ -22,7 +22,7 @@ public class WebAssemblyLocalizationTest : ServerTestBase<ToggleExecutionModeSer
         }
 
-        [Theory]
+        [Theory(Skip = "https://github.com/dotnet/runtime/issues/38124")]
```
Localization doesn't work

pranavkm commented on src/Components/test/E2ETest/Tests/WebAssemblyLoggingTest.cs:
```diff
@@ -52,7 +52,8 @@ public void LogsInnerExceptionsUsingLogger()
                 "[Custom logger] Unhandled exception rendering component: Here is the outer exception",
-                "System.InvalidTimeZoneException: Here is the outer exception ---> System.ArithmeticException: Here is the inner exception",
+                "System.InvalidTimeZoneException: Here is the outer exception",
+                "System.ArithmeticException: Here is the inner exception",
```
Exception message format is different

SteveSandersonMS replied:
Nae bother

pranavkm commented on src/ProjectTemplates/Shared/Project.cs:
```diff
@@ -117,9 +117,7 @@ internal async Task<ProcessResult> RunDotNetPublishAsync(IDictionary<string, str
             // Avoid restoring as part of build or publish. These projects should have already restored as part of running dotnet new. Explicitly disabling restore
             // should avoid any global contention and we can execute a build or publish in a lock-free way
 
-            var razorSDKarg = string.IsNullOrEmpty(RazorSdkDirectoryRoot) ? string.Empty : $"/p:RazorSdkDirectoryRoot={RazorSdkDirectoryRoot}";
```
This gets set via the props file

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/InitializeTestProjectAttribute.cs:
```diff
@@ -36,12 +36,7 @@ public override void Before(MethodInfo methodUnderTest)
 
-            MSBuildIntegrationTestBase.Project = ProjectDirectory.Create(_originalProjectName, _testProjectName, _baseDirectory, _additionalProjects, _language);
-#if NETCOREAPP
-            MSBuildIntegrationTestBase.TargetFramework = "net5.0";
```
I updated the test infrastructure to match the improvements we made in blazor.build

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Wasm/WasmPublishIntegrationTest.cs:
```diff
@@ -61,31 +59,28 @@ public async Task Publish_WithDefaultSettings_Works()
                 assetsManifestPath: "custom-service-worker-assets.js");
-
-            // When publishing without linker, we expect to have collation information present in mscorlib.dll
-            Assert.AssemblyContainsResource(result, Path.Combine(blazorPublishDirectory, "_framework", "_bin", "mscorlib.dll"), "collation.cjkCHS.bin");
```
We have an issue - https://github.com/dotnet/aspnetcore/issues/22983 - to figure out if there's a replacement for this if any.

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Wasm/WasmPublishIntegrationTest.cs:
```diff
@@ -61,31 +59,28 @@ public async Task Publish_WithDefaultSettings_Works()
             var blazorPublishDirectory = Path.Combine(publishDirectory, "wwwroot");
 
             Assert.FileExists(result, blazorPublishDirectory, "_framework", "blazor.boot.json");
-            Assert.FileExists(result, blazorPublishDirectory, "_framework", "blazor.webassembly.js");
```
I should bring this back.

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Wasm/WasmPublishIntegrationTest.cs:
```diff
@@ -555,102 +599,43 @@ private static void VerifyBootManifestHashes(MSBuildResult result, string blazor
             }
-        }
-
-        private static void VerifyBootManifestHashes(MSBuildResult result, string blazorPublishDirectory, ResourceHashesByNameDictionary resources, Func<string, string> relativePathFunc)
```
This got moved to a common file.

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/CreateBlazorTrimmerRootDescriptorFile.cs:
```diff
@@ -3,42 +3,53 @@
-            foreach (var assemblyName in assemblyNames)
+            foreach (var assembly in Assemblies)
             {
+                var assemblyName = assembly.GetMetadata("FileName") + assembly.GetMetadata("Extension");
```
We should be able to remove these trimmer files once https://github.com/dotnet/sdk/issues/12035 lands.

SteveSandersonMS replied:
https://github.com/dotnet/sdk/issues/12035 has a lot of ongoing discussion about what is and isn't in scope, so it's hard for me to be sure what it is/isn't going to do for us in the end.

With the implementation you have here, it looks like you're retaining the type-granularity linking we did before. Is that correct? If so that's great - we may well end up needing to keep this logic.

pranavkm replied:
Yup, we have the type level granularity from before. They are aware of what we are doing, so my hope if we get something cheap.

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/GenerateBlazorWebAssemblyBootJson.cs:
```diff
@@ -0,0 +1,142 @@
+                    var assetType = resource.GetMetadata("AssetType");
+                    var resourceName = $"{fileName}{extension}";
+
+                    if (!string.IsNullOrEmpty(resourceCulture))
```
We generate a different boot.json file for build and publish since the inputs are different. Calculating it in the task avoids duplicating this code in msbuild targets and unit testable.

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/GenerateServiceWorkerAssetsManifest.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace Microsoft.AspNetCore.Razor.Tasks
+{
+    public partial class GenerateServiceWorkerAssetsManifest : Task
```
Similar to boot.json, we operate on a different set of inputs during build and publish. This task has been modified to do what was previously done across multiple msbuild targets. I need to spend a little bit of time to verify, but it should be just as incremental as before.

SteveSandersonMS replied:
That's great. Do you feel the test coverage we have for this is adequate? We've had multiple regressions related to SWAM and build/publish/incrementalism, perhaps because it's pretty hard to tell whether it's right just by casually running an app.

pranavkm replied:
I retained all the tests from before. My plan's to go over these again and see if there are gaps and see if we could have more combination of tests (VS-style publish + build incrementalism for instance)

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.CodeGeneration.targets:
```diff
@@ -38,8 +38,6 @@ Copyright (c) .NET Foundation. All rights reserved.
     <_RazorGenerateInputsHash></_RazorGenerateInputsHash>
     <_RazorGenerateInputsHashFile>$(IntermediateOutputPath)$(MSBuildProjectName).RazorCoreGenerate.cache</_RazorGenerateInputsHashFile>
-
-    <_RazorToolAssembly Condition="'$(_RazorToolAssembly)'==''">$(RazorSdkDirectoryRoot)tools\netcoreapp3.0\rzc.dll</_RazorToolAssembly>
```
This got moved to Sdk.CurrentVersion.targets

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.StaticWebAssets.targets:
```diff
@@ -513,7 +513,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       <_ExternalPublishStaticWebAsset
         Include="%(StaticWebAsset.FullPath)"
-        Condition="'%(SourceType)' != ''">
+        Condition="'%(SourceType)' != '' AND '%(StaticWebAsset.CopyToPublishDirectory)' != 'Never'">
```
We don't want assets such as the boot.json produced during build to be copied to the publish directory. There was a bug in these targets where it wouldn't honor the CopyToPublishDirectory setting.

javiercn replied:
We shouldn't leave random stuff in the SWA definition. What problem are we trying to solve here?

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.props:
```diff
@@ -14,4 +14,5 @@ Copyright (c) .NET Foundation. All rights reserved.
   </PropertyGroup>
-</Project>
+
+</Project>
```
Will undo

pranavkm commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets:
```diff
@@ -354,6 +356,8 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <Import Project="Microsoft.NET.Sdk.Razor.MvcApplicationPartsDiscovery.targets" Condition="'$(_TargetingNETCoreApp30OrLater)' == 'true'" />
 
+  <Import Project="Microsoft.NET.Sdk.Razor.Components.Wasm.targets" Condition="'$(_TargetingNET50OrLater)' == 'true' AND '$(UseBlazorWebAssembly)' == 'true'" />
```
This target gets imported far too late to set up useful defaults such as `RuntimeIdentifier` or the flag that tells the SDK not to append the rid to the output path. We are able to do other things such as tell it to not reference M.AspNetCore.App or do trimmed publish. One way to remedy the evaluation order issue is to add a blazor-wasm specific SDK.

pranavkm committed:
Fixup

pranavkm committed:
Update js files

pranavkm committed:
Fixups

pranavkm committed:
Fixup

pranavkm committed:
Fixup

pranavkm committed:
Skip globalization tests

Pilchie commented:
:eyes:

pranavkm committed:
Undo workarounds

pranavkm commented on eng/Versions.props:
```diff
@@ -66,69 +66,69 @@
     <!-- Packages from dotnet/roslyn -->
     <MicrosoftNetCompilersToolsetPackageVersion>3.7.0-4.20319.6</MicrosoftNetCompilersToolsetPackageVersion>
     <!-- Packages from dotnet/runtime -->
-    <MicrosoftExtensionsDependencyModelPackageVersion>5.0.0-preview.7.20321.2</MicrosoftExtensionsDependencyModelPackageVersion>
```
Please ignore. I'll rebase this once darc updates this file. Using this to verify that the latest build resolves outstanding issues with the wasm runtime.

dougbu commented:
@pranavkm I got added as a reviewer due to code ownership I believe. And, you're going to revert the changes to the file I kinda own. Is there anything here you'd like me to review?

Curiosity only: Does this help move us any closer to archiving the dotnet/blazor repo?

SteveSandersonMS commented on src/Components/WebAssembly/testassets/StandaloneApp/StandaloneApp.csproj:
```diff
@@ -1,15 +1,17 @@
     <Reference Include="Microsoft.AspNetCore.Components.WebAssembly" />
     <Reference Include="System.Net.Http.Json" />
+
+    <FrameworkReference Remove="Microsoft.AspNetCore.App" />
```
Could you explain what this `<FrameworkReference>` is for? Since `StandaloneApp` runs on WebAssembly, I wouldn't expect there to be references to anything from server-side ASP.NET Core.

SteveSandersonMS commented on src/Components/test/E2ETest/Tests/WebAssemblyGlobalizationTest.cs:
```diff
@@ -10,11 +10,13 @@
     // culture in client-side Blazor.
-    public class WebAssemblyGlobalizationTest : GlobalizationTest<ToggleExecutionModeServerFixture<Program>>
+    // This type is internal since localization currently does not work.
+    // Make it public onc https://github.com/dotnet/runtime/issues/38124 is resolved.
```
Is there an issue tracking the need for us to do this? If not we might not remember to re-enable this test even when https://github.com/dotnet/runtime/issues/38124 is resolved

SteveSandersonMS commented on src/Components/test/E2ETest/Tests/WebAssemblyStringComparisonTest.cs:
```diff
@@ -21,7 +21,7 @@ public class WebAssemblyStringComparisonTest : ServerTestBase<ToggleExecutionMod
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/runtime/issues/38126")]
```
Would also be good to have an issue against Blazor for this to be re-enabled, unless it's tracked some other way.

SteveSandersonMS commented on src/Components/test/testassets/BasicTestApp/Program.cs:
```diff
@@ -91,6 +91,8 @@ private static async Task SimulateErrorsIfNeededForTest()
                 throw new InvalidTimeZoneException("This is an asynchronous startup exception");
             }
+
+            await Task.CompletedTask;
```
Why add this?

SteveSandersonMS commented on src/Components/test/testassets/BasicTestApp/wwwroot/js/jsinteroptests.js:
```diff
@@ -81,6 +81,7 @@ async function invokeDotNetInteropMethodsAsync(shouldSupportSyncInterop, dotNetO
   results['syncGenericInstanceMethod'] = await genericDotNetObjectByRef.invokeMethodAsync('Update', 'Updated value 1');
   results['asyncGenericInstanceMethod'] = await genericDotNetObjectByRef.invokeMethodAsync('UpdateAsync', 'Updated value 2');
 
+
```
Nit: revert

SteveSandersonMS commented on src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:
```diff
@@ -321,7 +321,7 @@ private async Task BlazorWasmHostedTemplate_IndividualAuth_Works(bool useLocalDb
 
                 aspNetProcess.VisitInBrowser(Browser);
-                TestBasicNavigation(project.ProjectName, usesAuth: true);
+                TestBasicNavigation(project.ProjectName, usesAuth: false /* Todo: https://github.com/dotnet/runtime/issues/38154 */);
```
Also let's be sure we're tracking this

pranavkm replied:
Missed undoing this. Fixed

SteveSandersonMS commented on src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:
```diff
@@ -536,6 +534,8 @@ private void TestBasicNavigation(string appName, bool usesAuth = false, bool ski
             }
 
+            // Skipping to workaround https://github.com/dotnet/runtime/issues/38154
+            skipFetchData = true;
```
Also, please either revert this (if it's fixed now) or add to the list of things we're tracking

SteveSandersonMS commented on src/Razor/Microsoft.AspNetCore.Razor.Tools/src/BrotliCompressCommand.cs:
```diff
@@ -0,0 +1,70 @@
+        public BrotliCompressCommand(Application parent)
+            : base(parent, "brotli")
+        {
+            Sources = Option("-s", ".cshtml files to compile", CommandOptionType.MultipleValue);
```
Is the text here a copy-paste mistake? We're not compressing `.cshtml` files...

SteveSandersonMS commented on src/Razor/Microsoft.AspNetCore.Razor.Tools/src/BrotliCompressCommand.cs:
```diff
@@ -0,0 +1,70 @@
+        {
+            Sources = Option("-s", ".cshtml files to compile", CommandOptionType.MultipleValue);
+            Outputs = Option("-o", "Generated output file path", CommandOptionType.MultipleValue);
+            CompressionLevelOption = Option("-c", "Compression level", CommandOptionType.SingleValue);
```
I don't really feel strongly about removing this, but I'm not aware we have any use case for having an option. We always want to use max AFAIK.

pranavkm replied:
They get turned down in the Razor integration tests. It makes the tests take far too long if they perform maximal compression every time.

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Wasm/WasmBuildIntegrationTest.cs:
```diff
@@ -0,0 +1,242 @@
+            Assert.Null(bootJsonData.resources.satelliteResources);
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/aspnetcore/issues/22975")]
```
To be tracked

pranavkm replied:
This one's our issue.

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Wasm/WasmBuildIntegrationTest.cs:
```diff
@@ -0,0 +1,242 @@
+        }
+
+        [Fact]
+        [QuarantinedTest]
```
Since so much has changed, would it be fair to unquarantine this as part of the update? Or do you think it's likely still to have issues?

javiercn replied:
We have a process in place for removing tests from quarantine. They need to be passing without failures for 30 days, so lets just follow the process to avoid disrupting the build

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/src/RazorGenerate.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Diagnostics;
```
Can be reverted?

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/LinkerWorkaround.xml:
```diff
@@ -19,6 +13,8 @@
 
+  <assembly fullname="System.Runtime.InteropServices.JavaScript" preserve="all" />
+
   <assembly fullname="System.Text.Json">
```
We *might* be able to drop this now (as it was working around an issue in STJ that might be fixed). Perfectly fine if you don't want to mess with extra stuff in this PR of course though!

pranavkm replied:
I filed https://github.com/dotnet/aspnetcore/issues/23262 to see if we can get these resolved by the BCL.

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Components.ServiceWorkerAssetsManifest.targets:
```diff
@@ -0,0 +1,166 @@
+
+WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and have
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
```
Wow, this all feels so official and grown-up now...

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Components.Wasm.targets:
```diff
@@ -0,0 +1,482 @@
+<!--
```
Since there's no diff associated with this file, it's pretty hard to CR and reason about what's changing.

Do you think you'd be able to summarize the high-level points about what's being done differently here now?

pranavkm replied:
The "Build" part of the pipeline is pretty close to what we had in Blazor.MonoRuntime.targets. We gather inputs, generate a blazor.boot.json, optionally a server worker manifest, add things to StaticWebAssets, and copy files to `$(OutputPath)wwwroot`. We no longer run the linker or compression during, so those parts of the pipeline are removed.

When publishing standalone, the linker operates on `ResolvedFileToPublish`. These files are inputs to compression, published PWA manifest, and the published blazor.boot.json. It's pretty much like build, but with different inputs. To make our lives easier, rather than the targets calculating values (such as `BootManifestResourceType`), the tasks do it. This saves us from duplicating a lot of MSBuild code between the build and publish targets.

 Publishing P2P is a bit more interesting. A `dotnet publish` on a hosted server project, only results in a `dotnet build` on the referenced client project. This is how MSBuild works. But we want this to produce a "published" WASM app. There is an escape hatch though. MSBuild also calls `GetCopyToPublishDirectoryItems` on referenced projects to get files to copy to the server's outputs. We use this to do a publish on the client project and copy them over to the server.

SteveSandersonMS commented on src/Razor/Microsoft.NET.Sdk.Razor/test/GenerateBlazorBootJsonTest.cs:
```diff
@@ -7,29 +7,28 @@
+    public class GenerateBlazorWebAssemblyBootJsonTest
     {
-        [Fact]
+        [Fact(Skip = "TODO fix this")]
```
Not sure if you're planning to handle this as part of the PR or later. If later, could you add it to the list of things to be tracked for follow-up? Presumably we could have one big issue listing them all.

SteveSandersonMS commented on src/Razor/test/testassets/Directory.Build.props:
```diff
@@ -26,6 +27,9 @@
     <AspNetCoreHostingModel>OutOfProcess</AspNetCoreHostingModel>
+
+    <!-- Turn down the compression level for brotli -->
+    <_BlazorBrotliCompressionLevel>NoCompression</_BlazorBrotliCompressionLevel>
```
Why? To make builds faster? Does it make enough of a difference to be worthwhile?

Also, our benchmarking relies on getting realistic brotli-compressed sizes. Just want to check this isn't going to interfere.

pranavkm replied:
This is specifically for the integration tests. It takes kinda long: https://github.com/dotnet/aspnetcore/issues/21382

SteveSandersonMS reviewed (commented):
This is looking really great! Thanks for figuring out all these details. It's cool to see things getting simpler as part of this process.

Not sure how close you are to considering this PR done so I'll await a further ping when you're ready for sign-off.

javiercn commented on src/Components/WebAssembly/testassets/HostedInAspNet.Client/HostedInAspNet.Client.csproj:
```diff
@@ -1,12 +1,10 @@
-    <!-- Disable compression in this project so that we can validate that it can be disabled -->
-    <BlazorEnableCompression>false</BlazorEnableCompression>
+    <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
+    <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
```
Did we get around the publish hosted bit by forcing the RID?

pranavkm replied:
I thought I'd tried this, but there's an issue where `dotnet publish -r [rid]` doesn't work. The restore that runs as part of publish messes things up. I'll talk to SDK folks about this and see what they recommend doing here.

javiercn commented on src/ProjectTemplates/Shared/Project.cs:
```diff
@@ -132,9 +130,7 @@ internal async Task<ProcessResult> RunDotNetBuildAsync(IDictionary<string, strin
-            var razorSDKarg = string.IsNullOrEmpty(RazorSdkDirectoryRoot) ? string.Empty : $"/p:RazorSdkDirectoryRoot={RazorSdkDirectoryRoot}";
-
-            using var result = ProcessEx.Run(Output, TemplateOutputDir, DotNetMuxer.MuxerPathOrDefault(), $"build --no-restore -c Debug /bl {razorSDKarg} {additionalArgs}", packageOptions);
+            using var result = ProcessEx.Run(Output, TemplateOutputDir, DotNetMuxer.MuxerPathOrDefault(), $"build --no-restore -c Debug /bl {additionalArgs}", packageOptions);
```
Why do we need the rid now here?

pranavkm replied:
Further up? WASM produces outputs in a RID based directory. There's a flag to turn it off, but does not work: https://github.com/dotnet/sdk/issues/12114

javiercn commented on src/Razor/Microsoft.AspNetCore.Razor.Tools/src/Application.cs:
```diff
@@ -39,6 +39,7 @@ internal class Application : CommandLineApplication
             Commands.Add(new ShutdownCommand(this));
             Commands.Add(new DiscoverCommand(this));
             Commands.Add(new GenerateCommand(this));
+            Commands.Add(new BrotliCompressCommand(this));
```
Does this ties compression to the Razor compiler?

pranavkm replied:
To the rzc tool, but it does not remote calls to the server process. It doesn't seem problematic as yet, but we can come back to this if there are side-effects to it.

javiercn commented on src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Microsoft.NET.Sdk.Razor.StaticWebAssets.targets:
```diff
@@ -269,7 +269,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       <StaticWebAsset
         Include="@(_ReferencedProjectStaticWebAssets)"
-        KeepMetadata="ContentRoot;BasePath;RelativePath;SourceId;SourceType" />
+        KeepMetadata="ContentRoot;BasePath;RelativePath;SourceId;SourceType;CopyToPublishDirectory" />
```
SWAs should only have `ContentRoot;BasePath;RelativePath;SourceId;SourceType` I want to avoid them becoming a random bag of properties.

javiercn reviewed (commented):
Looks great!

I want us to discuss the issue with SWA, but other than that looks fine.

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/runtime build 20200622.6

System.ComponentModel.Annotations , System.Diagnostics.DiagnosticSource , System.Diagnostics.EventLog , Microsoft.Extensions.Logging.Configuration , Microsoft.Extensions.Logging.Console , Microsoft.Extensions.Logging.Debug , Microsoft.Extensions.Logging.EventLog , Microsoft.Extensions.Logging.EventSource , Microsoft.Extensions.Logging.TraceSource , Microsoft.Extensions.Options , Microsoft.Extensions.Options.ConfigurationExtensions , Microsoft.Extensions.Options.DataAnnotations , Microsoft.Extensions.Primitives , Microsoft.Extensions.Logging.Abstractions , Microsoft.Extensions.Logging , Microsoft.Extensions.Internal.Transport , Microsoft.Extensions.Http , Microsoft.Extensions.Caching.Abstractions , Microsoft.Extensions.Caching.Memory , Microsoft.Extensions.Configuration , Microsoft.Extensions.Configuration.Abstractions , Microsoft.Extensions.Configuration.Binder , Microsoft.Extensions.Configuration.CommandLine , Microsoft.Extensions.Configuration.EnvironmentVariables , Microsoft.Extensions.Configuration.FileExtensions , Microsoft.Extensions.Configuration.Ini , Microsoft.Extensions.Configuration.Json , Microsoft.Extensions.Configuration.UserSecrets , Microsoft.Extensions.DependencyInjection , Microsoft.Extensions.DependencyInjection.Abstractions , Microsoft.Extensions.DependencyModel , Microsoft.Extensions.FileProviders.Abstractions , Microsoft.Extensions.FileProviders.Composite , Microsoft.Extensions.FileProviders.Physical , Microsoft.Extensions.FileSystemGlobbing , Microsoft.Extensions.HostFactoryResolver.Sources , Microsoft.Extensions.Hosting , Microsoft.Extensions.Hosting.Abstractions , Microsoft.Extensions.Configuration.Xml , Microsoft.NETCore.App.Runtime.win-x64 , Microsoft.NETCore.Platforms , Microsoft.Win32.Registry , Microsoft.Win32.SystemEvents , Microsoft.NETCore.App.Internal , Microsoft.NETCore.App.Ref , System.Drawing.Common , System.ServiceProcess.ServiceController , System.Text.Encodings.Web , System.Text.Json , System.Threading.Channels , System.Windows.Extensions , System.Security.Principal.Windows , System.Security.Permissions , System.Security.Cryptography.Xml , System.IO.Pipelines , System.Net.Http.Json , System.Net.Http.WinHttpHandler , System.Net.WebSockets.WebSocketProtocol , System.Reflection.Metadata , System.Runtime.CompilerServices.Unsafe , System.Security.Cryptography.Cng , System.Security.Cryptography.Pkcs
 From Version 5.0.0-preview.7.20321.2 -> To Version 5.0.0-preview.7.20322.6

pranavkm committed:
PR feedback

pranavkm committed:
Update js file

pranavkm committed:
Fix version

pranavkm committed:
Fixup

pranavkm committed:
Merge branch 'release/5.0-preview7' into prkrishn/net5-sdk

mkArtakMSFT reviewed (approved):
Let's get this merged so we can unblock our partners.
Please address any follow-up feedback through separate PRs, as this is pretty large already.
